### PR TITLE
Start refactor

### PR DESCRIPTION
**Pull request by klahnakoski**
pyparsing can be refactored while maintaining backward capability.  Here is a small refactor as a proof-of-concept.  I contend that moving code out into modules will make it easier to navigate, and make the overall architecture easier to learn.

If this PR is accepted, then I will propose more code be moved out to logical modules; eventually reducing `pyparsing.__init__.py` to the single function of declaring the namespace (the `__all__` export).

I am uncertain about what tests (and examples) must be run to confirm correctness, so I used the `.travis.yml` file for inspiration:

```
SET PYTHONPATH=.
c:\python36\python.exe simple_unit_tests.py
c:\python36\python.exe unitTests.py
c:\python36\python.exe examples/numerics.py
c:\python36\python.exe examples/TAP.py
c:\python36\python.exe examples/romanNumerals.py
c:\python36\python.exe examples/sexpParser.py
c:\python36\python.exe examples/oc.py
c:\python36\python.exe examples/delta_time.py
c:\python36\python.exe examples/eval_arith.py
```

### CODE REVIEW

**Comment by klahnakoski** [comment]
see https://github.com/pyparsing/pyparsing/issues/66

**Comment by cngkaygusuz** [comment]
I second this. There is more to be refactored, but this is a good start.

**Comment by klahnakoski** [comment]
@cngkaygusuz Yes, there is a lot more that can be moved out of the main source file. 

I think the biggest concern with refactoring is maintaining backward compatibility. Moving code around runs the risk of breaking other's pyparsing import statements. This refactor shows code can move while not changing the public/exported namespace; Specifically, we import all names to `pyparsing.__init__.py` and export (via __all__) from there.  Users will see no change.

**Comment by cngkaygusuz** [comment]
`pyparsing.py` defines `__all__`. Since the module was explicit in what it exposes to the outside world, it should be trivial to do the factoring while breaking no code.

I think Paul's biggest objection to the refactoring effort is the "drop-in" feature, a la simply copypasting `pyparsing.py` to an arbitrary location in a project and use it that way. I saw he was talking about it on Twist but I couldn't find it again, so I'll present my arguments why we must do this, here.

Converting `pyparsing.py` into a directory still keeps the drop-in functionality, instead of copying a file, a user merely has to copy a folder instead. I recall "inability to use a package manager" was mentioned, but it has more to do with pyparsing's dependencies (which we have none at the moment) rather than code organization. From a user PoV, I can't think of anything that would be different for them. 

We can keep the single file approach for a little while longer; however, since the development of the pyparsing is now more open, I expect the pace to pick up, and the line count will grow. Staying on single file won't be scalable, and it makes merge conflicts much more likely. It is better to solve this problem now than in the future.

In summary, this PR is a nudge towards best practices while having no downsides (that I can think of).

**Comment by klahnakoski** [comment]
If this refactor is accepted, then here is the next one (still a work on progress) https://github.com/klahnakoski/pyparsing/pull/1

**Comment by ptmcg** [comment]
Any chance we could get some input from projects that have included pyparsing in their vendored directories?

**Comment by klahnakoski** [comment]
@ptmcg 

1. The refactoring could be presented as a new major version number (as a release candidate `3.0.0.rc1` or maybe just `3.0.0`) or
2. The refactoring can proceed on a branch, with periodic releases under some other name (`pyparsing3`) until there is confidence the new branch works as expected.

**Comment by ptmcg** [comment]
Dripping water wears through stone. Given that we are doing a major release for dropping Py2 support, I guess we might as well do this breakup of the source code as well. I agree with Kyle that this should be doable without any compatibility concerns, but even so, a major release is the perfect time for such a radical restructuring. I remain unconvinced that dropping in a directory is as easy as dropping in a single source file, but it seems I am the only one who really cares about that capability.

I'll give this PR a closer read and see how things sort out.

Thanks for your patience and perseverance.

-- Paul

**Comment by ptmcg** [comment]
First impression: wow, that is a lot of code in `__init__.py`! I envision it to be more like a collection of imports from other modules. I may take a look at some projects like requests and Flask to see what their structure looks like.  Also, this project looks very interesting, from both a structure standardization point, and from a long-term automation point: https://chriswarrick.com/blog/2017/04/09/python-project-template-structure-and-automation/

**Comment by klahnakoski** [comment]
@ptmcg This PR is only a proof of concept; the concept that a directory can replace a single file.  I also wanted to make minimal change.  

If you are serious about breaking pyparsing into files, then I can continue working on this.  Here is an example branch I was working on: https://github.com/klahnakoski/pyparsing/blob/refactor2/pyparsing/__init__.py

Now that I am familiar with the code, I can do a little bit better.

**Comment by klahnakoski** [comment]
@ptmcg My biggest concern is the apparent lack of tools that would help you verify the refactoring; specifically, verifying that nothing important changed when code is moved from one file to another.  Github only shows a large number of lines removed, and large number of lines added elsewhere; which obscures any changes that could have been made.  For example, [I cut-and-pasted code into multiple files](https://github.com/klahnakoski/pyparsing/commit/80fe7d119d438909083a6cece2f0e761aea2e931), and it is a hard task to confirm there is not code change.

**Comment by ptmcg** [comment]
Shouldn't our unit tests give us some reassurance that we haven't left out anything important? I think we are at about 83% coverage.

**Comment by ptmcg** [comment]
@klahnakoski  I invited you onto the pyparsing-dev team so we could have some discussions there before jumping to code changes and PRs. For instance, I'm curious about your work in your refactor2 branch removing the dict from ParseResults - what motifvated this? This kind of thing would be great to discuss ahead of time, before making lots of code changes.

**Comment by klahnakoski** [comment]
@ptmcg I assume you mean the pyparsing-dev team on twist?

There are a few branches over at my fork, all of them are experimental:  The `refactor2` branch was an initial attempt to organize the project into multiple modules.  I am not convinced the current lines of separation are the best. In any event, the refactor2 branch is just about how to handle the inter-dependencies:  You may notice that at the top of some modules lines like:

https://github.com/klahnakoski/pyparsing/blob/4cba0493a4fd254670e42377d6ac48302272ff40/pyparsing/parser/base.py#L13

```python
SkipTo, ZeroOrMore, OneOrMore, Optional, NotAny, Suppress, _flatten, replaceWith, quotedString, And, MatchFirst, Or, Each, Empty, StringEnd, Literal, Token = [None] * 17
```

and at the bottom of others, lines like 

https://github.com/klahnakoski/pyparsing/blob/4cba0493a4fd254670e42377d6ac48302272ff40/pyparsing/parser/helpers.py#L1378

```python
# export
from pyparsing.parser import base

base._flatten = _flatten
base.replaceWith = replaceWith
base.quotedString = quotedString
```

This refactoring did not take long, since it was just cut-and-paste.  So re-doing it correctly would not be difficult. Yes, I would appreciate discussion before I try again.  I also see I moved tests around, but I am willing to re-do that too.

The `remove-dict` branch is an experiment: It is an attempt to make the `ParseResults` much smaller, and to standardize the various parser output to a single datatype.  The hope is smaller means faster because so many are generated. The longer I spend on it, the less I believe I will ever get working.  At least I understand the code better now.

**Comment by ptmcg** [comment]
@klahnakoski - no I was referring to the pyparsing-dev team I started here on GitHub. Twist just turns out to be just one more place I need to look for comments and discussion. While the GitHub team discussions don't seem to be too tightly tied to the repos, I hold out hope that this can be figured out given the shared platform. So I invited you to the GitHub pyparsing-dev team, and there is already a discussion thread on doing this refactor.  Your post prompts some more questions in my head, but I'll post them there for the benefit of all the team members, and for our glorious posterity.

**Comment by klahnakoski** [comment]
@ptmcg maybe I need permissions, or a link, to view the "pyparsing-dev team".  I can not find it.

**Comment by ptmcg** [comment]
I just re-sent the invitation, or go to  https://github.com/pyparsing (organization link, not the repo link).

**Comment by klahnakoski** [comment]
Closing this, it was meant to start a discussion, and its job is done
